### ISSUE TITLE
knitr::purl fail with parametrized Rmd when params is used in chunk option

### ISSUE THREAD

**Issue by cderv** [yihui/knitr#1938]
From https://github.com/wlandau/targets/issues/256

When `tangle = TRUE` and Rmd uses parameters, it works ok if params are used in code chunks but it fails if params is used in one of the chunk option. 

``` r
lines <- c(
  "---",
  "title: parameters",
  "output: html_document",
  "params:",
  "  paramcd: TRUE",
  "---",
  "",
  "```{r, eval = params$paramcd}",
  "print(params$paramcd)",
  "print(object)",
  "```"
)
writeLines(lines, "report.Rmd")
knitr::knit("report.Rmd", tangle = TRUE, quiet = TRUE)
#> Error in eval(x, envir = envir) : objet 'params' introuvable
#> [1] "report.R"
xfun::file_string("report.R")
#> params <-
#> list(paramcd = TRUE)

lines[8] <- "```{r, eval = TRUE}"
writeLines(lines, "report.Rmd")
knitr::knit("report.Rmd", tangle = TRUE, quiet = TRUE)
#> [1] "report.R"
xfun::file_string("report.R")
#> params <-
#> list(paramcd = TRUE)
#> 
#> ## ---- eval = TRUE-------------------------------------------------------------
#> print(params$paramcd)
#> print(object)
```

It seems like a bug. 

Also, it seems **knitr** is not aware about parameters defined in a Rmd file when knitting. But this may be unsupported as parametrized report is a **rmarkdown** feature mainly

```r
lines <- c(
  "---",
  "title: knitr with params",
  "output: html_document",
  "params:",
  "  paramcd: TRUE",
  "---",
  "",
  "```{r, eval = TRUE}",
  "str(params)",
  "```"
)
writeLines(lines, "report.Rmd")
res <- knitr::knit("report.Rmd", quiet = TRUE)
xfun::file_string(res)
#> ---
#> title: knitr with params
#> output: html_document
#> params:
#>   paramcd: TRUE
#> ---
#> 
#> ```r
#> str(params)
#> #> Error in str(params): object 'params' not found
#> ```
```

**Comment by yihui** [comment]
> But this may be unsupported as parameterized report is a **rmarkdown** feature mainly

That's it.

**Comment by cderv** [comment]
Then we can say it is a feature request to be able to "purl" Rmd file, even when using parameters ? 

It could also be another tool than **knitr**, as parsing code chunks and inline code from a Rmd chunk is something that seem to be often needed. For now, I think every usage like that uses `purl()`.

**Comment by yihui** [comment]
In principle, `purl()` doesn't evaluate any code in the document (only `knit()` does), so it won't recognize `params$paramcd`. To support this feature, we may have to break the design.

**Comment by cderv** [comment]
I see. Thanks! 

I think there is a need for Rmd code parsing with evaluation (as the one in **targets** - cc @wlandau @yonicd) but maybe it has to be different from `purl`. I'll keep think about it for the long term, looking at other existing use case.

**Comment by yihui** [comment]
I'm okay with supporting this case. I have already had some code supporting the special case of R Markdown in **knitr**. Parsing YAML doesn't involve "evaluating" code, so that's probably fine.

In fact, `params` is supported when you do not use it in chunk options: https://github.com/yihui/knitr/blob/0272a83dbb8f687ee1d2be7f3ee7c11908d5845e/R/output.R#L235-L238

The deeper problem is that I have never been satisfied with `purl()`. It is hard and sometimes impossible to do it right without evaluating code. The true way to purl code from a document is actually to use the hook `knitr::hook_purl()`.

**Comment by wlandau** [comment]
> The true way to purl code from a document is actually to use the hook knitr::hook_purl().

I think there is still a need to statically extract code from a `knitr` document from outside the document itself (examples from [`targets`](https://books.ropensci.org/targets/files#literate-programming) and [`drake`](https://books.ropensci.org/drake/plans.html#static-files)).

**Comment by yihui** [comment]
Yes, there is a need. The problem is that it is not possible to both statically and faithfully extract code in certain cases. The `hook_purl` is must more reliable than `purl()`, but the obvious disadvantage is that it requires knitting the document.

**Comment by yihui** [comment]
Fixed in a39a31130df777309ff0531a1e92feb7cf6b490b. When tangling (`purl()`) an R Markdown document with a `params` field in its YAML front matter, chunk options that reference those params (e.g. `eval = params$paramcd`) are now resolved, so the chunk is no longer silently dropped from the tangled script.

Your original reprex now produces:

```r
params <-
list(paramcd = TRUE)

## ----eval = params$paramcd----------------------------------------------------
print(params$paramcd)
print(object)
```

As you noted upthread, parsing the YAML `params` does not involve evaluating any document code, so tangling stays a purely static operation — the params list is just made available in the knitting environment while chunk options are evaluated. Thanks, @cderv!
